src/sentry/db/models/base.py

```diff
@@ -92,6 +94,55 @@ class Meta:
     __repr__ = sane_repr("id")
 
 
+@dataclass
+class SnowflakeBitSegment:
+    length: int
+    name: str
+
+    def validate(self, value):
+        if value >> self.length != 0:
+            raise Exception(f"{self.name} exceed max bit value of {self.length}")
+        return True
+
+
+class Snowflake:
+    SENTRY_EPOCH_START = datetime(2022, 4, 26, 0, 0).timestamp()
+    SNOWFLAKE_ID_LENGTH = SnowflakeBitSegment(53, "Snowflake ID")
+
+    SEGMENT_LENGTH = {
+        "VERSION_ID": SnowflakeBitSegment(5, "Version ID"),
+        "TIME_DIFFERENCE": SnowflakeBitSegment(32, "Time difference"),
+        "REGION_ID": SnowflakeBitSegment(12, "Region ID"),
+        "REGION_SEQUENCE": SnowflakeBitSegment(4, "Region sequence"),
+    }
+
+    SEGMENT_VALUE = {
+        "VERSION_ID": 0,
+        "TIME_DIFFERENCE": 0,
+        "REGION_ID": 0,
+        "REGION_SEQUENCE": 0,
+    }
+
+    id = BoundedBigAutoField(primary_key=True)
+
+    def snowflake_id_generation(self):
+        current_time = datetime.now().timestamp()
+        # supports up to 130 years
+        self.SEGMENT_VALUE["TIME_DIFFERENCE"] = int(current_time - self.SENTRY_EPOCH_START)
+
+        total_bits_to_allocate = 53
+        snowflake_id = 0
+
+        for key, segment in self.SEGMENT_LENGTH.items():
```

Comment:
@maxiuyuan `self.SEGMENT_LENGTH` needs to be an ordered dictionary so that `item()` is predictable.